testing/test_gflownet.py

```diff
@@ -0,0 +1,64 @@
+from gfn.containers import Trajectories
+from gfn.containers.base import Container
+from gfn.gflownet import FMGFlowNet, TBGFlowNet
+from gfn.gym import Box
+from gfn.gym.helpers.box_utils import (
+    BoxPBEstimator,
+    BoxPBNeuralNet,
+    BoxPFEstimator,
+    BoxPFNeuralNet,
+)
+from gfn.modules import DiscretePolicyEstimator
+from gfn.states import States
+
+
+def test_trajectory_based_gflownet_generic():
+    pf_module = BoxPFNeuralNet(
+        hidden_dim=32, n_hidden_layers=2, n_components=1, n_components_s0=1
+    )
+    pb_module = BoxPBNeuralNet(
+        hidden_dim=32, n_hidden_layers=2, n_components=1, torso=pf_module.torso
+    )
+
+    env = Box()
+
+    pf_estimator = BoxPFEstimator(
+        env=env, module=pf_module, n_components=1, n_components_s0=1
+    )
+    pb_estimator = BoxPBEstimator(env=env, module=pb_module, n_components=1)
+
+    gflownet = TBGFlowNet(pf=pf_estimator, pb=pb_estimator)
+    mock_trajectories = Trajectories(env)
+
+    result = gflownet.to_training_samples(mock_trajectories)
+
+    # Assert that the result is of type `Trajectories`
+    assert isinstance(
+        result, Container
+    ), f"Expected type Container, but got {type(result)}"
+    assert isinstance(
+        result, Trajectories
+    ), f"Expected type Trajectories, but got {type(result)}"
+
+
+def test_flow_matching_gflownet_generic():
+    env = Box()
+    module = BoxPFNeuralNet(
+        hidden_dim=32, n_hidden_layers=2, n_components=1, n_components_s0=1
+    )
+    estimator = DiscretePolicyEstimator(env, module, True)
+    gflownet = FMGFlowNet(estimator)
+    mock_trajectories = Trajectories(env)
+    states_tuple = gflownet.to_training_samples(mock_trajectories)
+
+    # Assert that the result is a tuple of `States`
+    assert isinstance(
+        states_tuple, tuple
+    ), f"Expected type tuple, but got {type(states_tuple)}"
+
+    assert isinstance(
```

Comment:
Thanks very much for this PR. I am leaning toward thinking this is a nice addition to the library, and I really appreciate the nice tests (though I'm curious what @saleml thinks as well).

I'm wondering two things:

1) could you please add another test that *asserts that an exception is raised* - I.e., when someone defined the wrong return type in a custom `GFlowNet` class definition, or the GFN returns the wrong type, or something else of this nature. That way we can be sure the library does the right thing when the developer makes a mistake.

2) can we add a short example somewhere (it need not be sensible) that demonstrates clearly what a user needs to do to ensure output types are handled properly?

Reply:
This is indeed a nice addition to the library! Thank you for the PR. I agree it would be nice to have a small example somewhere, maybe in the README ?

Reply:
@saleml are you thinking some documentation to show how to add a new implementation of a `GFlowNet`? If so, I'd be happy to add it to the README, but perhaps it might be better suited to go somewhere like `tutorials/GFLOWNET.md` or something? 

It seems like adding new `GFlowNet` subclasses will not be a super common developer use-case so it might be worth keeping the README to the developer happy paths? Also it seems like it might fit nicely alongside `tutorials/ENV.md` which has similar properties of being a less common, more core development oriented task involving expanding the core object model of the library.

That said, I am of course happy to defer to whatever you would prefer.

Reply:
I've taken a crack at adding this in `tutorials/GFLOWNET.md` and linked to it from the main README similarly to `tutorials/ENV.md`. Please feel free to let me know if you'd like more/less/different detail on any aspect of the doc and/or structural changes. 🙏